File: python/670.py
"""Project Euler Problem 670: Coloured Tiles.

Find the number of ways to fill a 2 x N rectangle with 1x1, 1x2, ... 1xT
tiles such that no four tiles meet at a single point, each tile is one of K
colors, and no adjacent tiles have the same color.

We use matrix multiplication with states representing tile configurations.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

from sympy import Matrix


@dataclass(frozen=True)
class Point:
    """Integer point."""

    x: int
    y: int


def mat_pow_mod(mat: list[list[int]], exp: int, mod: int) -> list[list[int]]:
    """Matrix power modulo mod."""
    n = len(mat)
    result = [[1 if i == j else 0 for j in range(n)] for i in range(n)]
    base = [[mat[i][j] % mod for j in range(n)] for i in range(n)]

    while exp > 0:
        if exp & 1:
            result = mat_mult(result, base, mod)
        base = mat_mult(base, base, mod)
        exp >>= 1

    return result


def mat_mult(
    a: list[list[int]], b: list[list[int]], mod: int
) -> list[list[int]]:
    """Matrix multiplication modulo mod."""
    n = len(a)
    result = [[0] * n for _ in range(n)]
    for i in range(n):
        for j in range(n):
            for k in range(n):
                result[i][j] = (result[i][j] + a[i][k] * b[k][j]) % mod
    return result


def solve() -> int:
    """Solve Problem 670."""
    N = 10**16
    K = 4
    T = 3
    M = 1000004321

    # Build states
    states: list[Point] = [Point(-1, -1)]
    for i in range(T):
        for j in range(T):
            states.append(Point(i, j))

    ordering: Dict[Point, int] = {p: i for i, p in enumerate(states)}

    # Build transition matrix
    A = [[0] * len(states) for _ in range(len(states))]
    A[0][0] = K - 1
    A[0][1] = K - 2

    for i in range(T):
        for j in range(T):
            A[ordering[Point(i, j)]][0] = (K - 1) * (K - 2)

    for i in range(1, T):
        for j in range(T):
            A[ordering[Point(i - 1, j)]][ordering[Point(i, 0)]] = K - 2

    for i in range(T):
        for j in range(1, T):
            A[ordering[Point(i, j - 1)]][ordering[Point(0, j)]] = K - 2

    for i in range(1, T):
        for j in range(1, T):
            A[ordering[Point(i - 1, j - 1)]][ordering[Point(i, j)]] = 1

    # Compute A^(N-1)
    Ae = mat_pow_mod(A, N - 1, M)

    ans = 0
    for t in range(2):
        ans = (ans + K * Ae[t][0]) % M
        for i in range(T):
            for j in range(T):
                ans = (
                    ans + K * (K - 1) * Ae[t][ordering[Point(i, j)]]
                ) % M

    return ans


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/670.c
/*
 * Project Euler 670 - Coloured Tiles
 *
 * Matrix exponentiation approach with states representing tile configs.
 * N=10^16, K=4, T=3, M=1000004321.
 */
#include <stdio.h>
#include <string.h>

typedef long long ll;
typedef __int128 lll;

#define MOD 1000004321LL
#define SZ 10  /* 1 + T*T = 1 + 9 = 10 states */
#define T 3
#define K 4

typedef struct {
    ll m[SZ][SZ];
} Mat;

static Mat mat_mult(Mat *a, Mat *b) {
    Mat result;
    memset(&result, 0, sizeof(result));
    for (int i = 0; i < SZ; i++) {
        for (int k = 0; k < SZ; k++) {
            if (a->m[i][k] == 0) continue;
            for (int j = 0; j < SZ; j++) {
                result.m[i][j] = (result.m[i][j] + (lll)a->m[i][k] * b->m[k][j]) % MOD;
            }
        }
    }
    return result;
}

static Mat mat_pow(Mat *mat, ll exp) {
    Mat result;
    memset(&result, 0, sizeof(result));
    for (int i = 0; i < SZ; i++) result.m[i][i] = 1;

    Mat base = *mat;
    while (exp > 0) {
        if (exp & 1) result = mat_mult(&result, &base);
        base = mat_mult(&base, &base);
        exp >>= 1;
    }
    return result;
}

/* States: 0 = (-1,-1), then (i,j) for i=0..T-1, j=0..T-1 mapped to 1+i*T+j */
static int point_idx(int i, int j) {
    return 1 + i * T + j;
}

int main() {
    ll N = 10000000000000000LL; /* 10^16 */

    Mat A;
    memset(&A, 0, sizeof(A));

    /* A[0][0] = K-1 = 3 */
    A.m[0][0] = K - 1;
    /* A[0][1] = K-2 = 2 (index of Point(0,0) = 1) */
    A.m[0][point_idx(0, 0)] = K - 2;

    for (int i = 0; i < T; i++) {
        for (int j = 0; j < T; j++) {
            A.m[point_idx(i, j)][0] = (ll)(K - 1) * (K - 2);
        }
    }

    for (int i = 1; i < T; i++) {
        for (int j = 0; j < T; j++) {
            A.m[point_idx(i - 1, j)][point_idx(i, 0)] = K - 2;
        }
    }

    for (int i = 0; i < T; i++) {
        for (int j = 1; j < T; j++) {
            A.m[point_idx(i, j - 1)][point_idx(0, j)] = K - 2;
        }
    }

    for (int i = 1; i < T; i++) {
        for (int j = 1; j < T; j++) {
            A.m[point_idx(i - 1, j - 1)][point_idx(i, j)] = 1;
        }
    }

    /* Compute A^(N-1) */
    Mat Ae = mat_pow(&A, N - 1);

    ll ans = 0;
    for (int t = 0; t < 2; t++) {
        ans = (ans + (ll)K * Ae.m[t][0]) % MOD;
        for (int i = 0; i < T; i++) {
            for (int j = 0; j < T; j++) {
                ans = (ans + (ll)K * (K - 1) % MOD * Ae.m[t][point_idx(i, j)]) % MOD;
            }
        }
    }

    printf("%lld\n", ans);
    return 0;
}
